Plugin fails with ES-1.4.0.Beta1

Build 1.3.2 of the plugin crashes ES-1.4.0.Beta1 with the error:
`1) No implementation for org.elasticsearch.http.HttpServerTransport was bound.`

The RestAction branch build 1.3.2 also breaks with
`1) NoSuchMethodError[org.elasticsearch.rest.BaseRestHandler.<init>(Lorg/elasticsearch/common/settings/Settings;Lorg/elasticsearch/client/Client;)V]
make: *** [start] Error 3`

--- a/README.md
+++ b/README.md
@@ -1,7 +1,7 @@
 
 **IMPORTANT NOTICE**: versions 1.0.4 and 1.1.0 are *insecure and should not be used*.
 They have a bug that allows an attacker to get ip authentication by setting
-its ip on the 'Host' header. A fix is provided for now for versions v1.2.0 and
+its ip on the 'Host' header. A fix is provided for versions v1.2.0 and
 v.1.3.0 of the plugin.
 
 # HTTP Basic auth for ElasticSearch
@@ -18,6 +18,7 @@ There is no way to configure this on a per index basis.
 
 |     Http Basic Plugin       | elasticsearch         |
 |-----------------------------|-----------------------|
+| v1.4.0.Beta1 (1.4.0.Beta1)  | 1.4.0.Beta1           |
 | v1.3.0(master)              | 1.3.0                 |
 | v1.2.0                      | 1.2.0                 |
 | 1.1.0                       | 1.0.0                 |

--- a/README.md
+++ b/README.md
@@ -18,8 +18,8 @@ There is no way to configure this on a per index basis.
 
 |     Http Basic Plugin       | elasticsearch         |
 |-----------------------------|-----------------------|
-| v1.4.0.Beta1 (1.4.0.Beta1)  | 1.4.0.Beta1           |
-| v1.3.0(master)              | 1.3.0                 |
+| v1.4.0(master)              | 1.4.0                 |
+| v1.3.0                      | 1.3.0                 |
 | v1.2.0                      | 1.2.0                 |
 | 1.1.0                       | 1.0.0                 |
 | 1.0.4                       | 0.90.7                |

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
 
     <groupId>com.asquera.elasticsearch</groupId>
     <artifactId>elasticsearch-http-basic</artifactId>
-    <version>1.3.0</version>
+    <version>1.4.0</version>
     <packaging>jar</packaging>
 
     <name>Basic Authentication Plugin</name>
@@ -12,8 +12,8 @@
 
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <elasticsearch.version>1.3.0</elasticsearch.version>
-        <lucene.version>4.9.0</lucene.version>
+        <elasticsearch.version>1.4.0.Beta1</elasticsearch.version>
+        <lucene.version>4.10.1</lucene.version>
     </properties>
 
     <dependencies>

--- a/pom.xml
+++ b/pom.xml
@@ -12,8 +12,9 @@
 
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <elasticsearch.version>1.4.0.Beta1</elasticsearch.version>
-        <lucene.version>4.10.1</lucene.version>
+        <elasticsearch.version>1.4.0</elasticsearch.version>
+        <lucene.version>4.10.2</lucene.version>
+        <lucene.maven.version>4.10.2</lucene.maven.version>
     </properties>
 
     <dependencies>
@@ -21,7 +22,7 @@
        <dependency>
           <groupId>org.apache.lucene</groupId>
           <artifactId>lucene-test-framework</artifactId>
-          <version>${lucene.version}</version>
+          <version>${lucene.maven.version}</version>
           <scope>test</scope>
         </dependency>
 
@@ -32,6 +33,12 @@
           <scope>test</scope>
         </dependency>
 
+        <dependency>
+          <groupId>com.carrotsearch.randomizedtesting</groupId>
+          <artifactId>randomizedtesting-runner</artifactId>
+          <version>2.1.10</version>
+          <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>org.elasticsearch</groupId>
             <artifactId>elasticsearch</artifactId>

--- a/src/main/java/com/asquera/elasticsearch/plugins/http/HttpBasicServerModule.java
+++ b/src/main/java/com/asquera/elasticsearch/plugins/http/HttpBasicServerModule.java
@@ -7,12 +7,13 @@ import org.elasticsearch.common.settings.Settings;
  * @author Florian Gilcher (florian.gilcher@asquera.de)
  */
 public class HttpBasicServerModule extends HttpServerModule {
-    
+
     public HttpBasicServerModule(Settings settings) {
         super(settings);
     }
-    
+
     @Override protected void configure() {
+        super.configure();
         bind(HttpBasicServer.class).asEagerSingleton();
     }
 }

--- a/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/DefaultConfigurationIntegrationTest.java
+++ b/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/DefaultConfigurationIntegrationTest.java
@@ -18,33 +18,24 @@
  */
 package com.asquera.elasticsearch.plugins.http.auth.integration;
 
-
-import org.apache.http.impl.client.HttpClients;
-import org.elasticsearch.common.settings.ImmutableSettings;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.rest.RestStatus;
-import org.elasticsearch.test.ElasticsearchIntegrationTest;
 import org.elasticsearch.test.ElasticsearchIntegrationTest.ClusterScope;
 import org.elasticsearch.test.ElasticsearchIntegrationTest.Scope;
-import org.elasticsearch.test.rest.client.http.HttpRequestBuilder;
 import org.elasticsearch.test.rest.client.http.HttpResponse;
 import org.junit.Test;
-
-import com.asquera.elasticsearch.plugins.http.HttpBasicServerPlugin;
 
 import static org.hamcrest.Matchers.equalTo;
 
 /**
  * Test a rest action that sets special response headers
  */
-@ClusterScope(transportClientRatio = 0.0, scope = Scope.SUITE, numDataNodes = 1)
-public class DefaultConfigurationIntegrationTest extends ElasticsearchIntegrationTest {
+@ClusterScope(scope = Scope.SUITE, numDataNodes = 1)
+public class DefaultConfigurationIntegrationTest extends HttpBasicServerPluginIntegrationTest {
 
     @Override
     protected Settings nodeSettings(int nodeOrdinal) {
-        return ImmutableSettings.settingsBuilder()
-                .put("plugin.types", HttpBasicServerPlugin.class.getName())
-                .build();
+      return builderWithPlugin().build();
     }
 
     @Test
@@ -54,12 +45,8 @@ public class DefaultConfigurationIntegrationTest extends ElasticsearchIntegratio
     }
 
     @Test
-    public void localhostClientIsAuthenticated() throws Exception {
+    public void localhostClientIsIpAuthenticated() throws Exception {
         HttpResponse response = httpClient().path("/_status").execute();
         assertThat(response.getStatusCode(), equalTo(RestStatus.OK.getStatus()));
-    }
-   
-    public static HttpRequestBuilder httpClient() {
-        return new HttpRequestBuilder(HttpClients.createDefault()).host("localhost").port(9200);
     }
 }

--- a/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/EmptyWhitelistIntegrationTest.java
+++ b/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/EmptyWhitelistIntegrationTest.java
@@ -18,25 +18,13 @@
  */
 package com.asquera.elasticsearch.plugins.http.auth.integration;
 
-import org.apache.http.client.methods.CloseableHttpResponse;
-import org.apache.http.client.methods.HttpUriRequest;
-import org.apache.http.impl.client.CloseableHttpClient;
-import org.apache.http.impl.client.HttpClients;
-import org.elasticsearch.common.settings.ImmutableSettings;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.Base64;
 import org.elasticsearch.rest.RestStatus;
-import org.elasticsearch.test.ElasticsearchIntegrationTest;
 import org.elasticsearch.test.ElasticsearchIntegrationTest.ClusterScope;
-import org.elasticsearch.test.rest.client.http.HttpGetWithEntity;
 import org.elasticsearch.test.rest.client.http.HttpRequestBuilder;
 import org.elasticsearch.test.rest.client.http.HttpResponse;
 import org.junit.Test;
-
-import com.asquera.elasticsearch.plugins.http.HttpBasicServerPlugin;
-
-import java.net.URI;
-import java.net.URISyntaxException;
 
 import static org.elasticsearch.test.ElasticsearchIntegrationTest.Scope;
 import static org.hamcrest.Matchers.equalTo;
@@ -45,71 +33,33 @@ import static org.hamcrest.Matchers.equalTo;
  * Test a rest action that sets special response headers
  */
 @ClusterScope(transportClientRatio = 0.0, scope = Scope.SUITE, numDataNodes = 1)
-public class EmptyWhitelistIntegrationTest extends ElasticsearchIntegrationTest {
+public class EmptyWhitelistIntegrationTest extends HttpBasicServerPluginIntegrationTest {
 
     @Override
     protected Settings nodeSettings(int nodeOrdinal) {
-        return ImmutableSettings.settingsBuilder().putArray("http.basic.ipwhitelist", "unkown")
-                 .put("plugin.types", HttpBasicServerPlugin.class.getName())
+        return  builderWithPlugin().
+                putArray("http.basic.ipwhitelist", "unkown")
                  .build();
     }
 
+// TODO put the set credentials ussing Setter
     @Test
-    public void testHealthCheck() throws Exception {
-        HttpResponse response = httpClient().path("/").execute();
+    public void clientIpAuthenticationFails() throws Exception {
+        HttpResponse response = httpClient().path("/_status").execute();
+        assertThat(response.getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
+    }
+
+    @Test
+    public void clientGoodCredentialsBasicAuthenticationSuceeds() throws Exception {
+        HttpResponse response = requestWithCredentials("admin:admin_pw")
+          .addHeader("X-Forwarded-For", "1.1.1.1" ).execute();
         assertThat(response.getStatusCode(), equalTo(RestStatus.OK.getStatus()));
     }
 
     @Test
-    public void localhostClientIsNotIpAuthenticated() throws Exception {
-        HttpResponse response = httpClient().path("/_status").execute();
-        assertThat(response.getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
+    public void clientBadCredentialsBasicAuthenticationFails() throws Exception {
+        HttpResponse response = requestWithCredentials("admin:wrong").execute();
+        assertThat(response.getStatusCode()
+            , equalTo(RestStatus.UNAUTHORIZED.getStatus()));
     }
-    
-    @Test
-    public void localhostClientIsBasicAuthenticated() throws Exception {
-        HttpUriRequest request = httpRequest();
-        String credentials = "admin:admin_pw";
-        request.setHeader("Authorization", "Basic " + Base64.encodeBytes(credentials.getBytes()));
-        CloseableHttpResponse response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.OK.getStatus()));
-    }
-    
-
-    @Test
-    public void localhostClientIsBasicAuthenticatedPassingXForward() throws Exception {
-        HttpUriRequest request = httpRequest();
-        String credentials = "admin:admin_pw";
-        request.setHeader("Authorization", "Basic " + Base64.encodeBytes(credentials.getBytes()));
-        request.setHeader("X-Forwarded-For", "1.1.1.1" );
-        CloseableHttpResponse response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.OK.getStatus()));
-    }
-    @Test
-    public void localhostClientNotBasicAuthenticated() throws Exception {
-        HttpUriRequest request = httpRequest();
-        String credentials = "admin:wrong";
-        request.setHeader("Authorization", "Basic " + Base64.encodeBytes(credentials.getBytes()));
-        CloseableHttpResponse response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
-    }
-
-    public static HttpRequestBuilder httpClient() {
-        return new HttpRequestBuilder(HttpClients.createDefault()).host("localhost").port(9200);
-    }
-   
-    public static HttpUriRequest httpRequest() {
-      HttpUriRequest httpUriRequest = null;
-          try {
-            httpUriRequest = new HttpGetWithEntity(new URI("http", null, "localhost", 9200, "/_status", null, null));
-                } catch (URISyntaxException e) {
-                  throw new IllegalArgumentException(e);
-                }
-      return httpUriRequest;
-    }
-
-    public static CloseableHttpClient closeableHttpClient() {
-      return HttpClients.createDefault();
-    }
-    
 }

--- a/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/HttpBasicServerPluginIntegrationTest.java
+++ b/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/HttpBasicServerPluginIntegrationTest.java
@@ -1,0 +1,49 @@
+
+package com.asquera.elasticsearch.plugins.http.auth.integration;
+
+
+import java.net.InetSocketAddress;
+
+import org.apache.http.impl.client.HttpClients;
+import org.elasticsearch.common.transport.InetSocketTransportAddress;
+import org.elasticsearch.common.Base64;
+import org.elasticsearch.http.HttpServerTransport;
+import org.elasticsearch.test.ElasticsearchIntegrationTest;
+import org.elasticsearch.test.rest.client.http.HttpRequestBuilder;
+import com.asquera.elasticsearch.plugins.http.HttpBasicServerPlugin;
+import org.elasticsearch.common.settings.ImmutableSettings.Builder;
+import org.elasticsearch.common.settings.ImmutableSettings;
+
+/**
+ *
+ * @author Ernesto Miguez (ernesto.miguez@asquera.de)
+ */
+
+public abstract class HttpBasicServerPluginIntegrationTest extends
+ElasticsearchIntegrationTest {
+
+  protected final String localhost = "127.0.0.1";
+
+
+  public static HttpRequestBuilder httpClient() {
+    HttpServerTransport httpServerTransport = internalCluster().getDataNodeInstance(HttpServerTransport.class);
+    InetSocketAddress address = ((InetSocketTransportAddress) httpServerTransport.boundAddress().publishAddress()).address();
+    return new HttpRequestBuilder(HttpClients.createDefault()).host(address.getHostName()).port(address.getPort());
+  }
+  /**
+   *
+   * @return a Builder with the plugin included and bind_host and publish_host
+   * set to localhost, from where the client's request ip will be done.
+   */
+  protected Builder builderWithPlugin() {
+    return ImmutableSettings.settingsBuilder()
+      .put("network.host", localhost)
+      .put("plugin.types", HttpBasicServerPlugin.class.getName());
+  }
+
+  protected HttpRequestBuilder requestWithCredentials(String credentials) throws Exception {
+        return httpClient().path("/_status")
+          .addHeader("Authorization", "Basic " + Base64.encodeBytes(credentials.getBytes()));
+    }
+
+}

--- a/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/IpAuthenticationIntegrationTest.java
+++ b/src/test/java/com/asquera/elasticsearch/plugins/http/auth/integration/IpAuthenticationIntegrationTest.java
@@ -1,4 +1,3 @@
-
 /*
  * Licensed to Elasticsearch under one or more contributor
  * license agreements. See the NOTICE file distributed with
@@ -19,25 +18,17 @@
  */
 package com.asquera.elasticsearch.plugins.http.auth.integration;
 
-import org.apache.http.client.methods.CloseableHttpResponse;
-import org.apache.http.client.methods.HttpUriRequest;
-import org.apache.http.impl.client.CloseableHttpClient;
-import org.apache.http.impl.client.HttpClients;
-import org.elasticsearch.common.settings.ImmutableSettings;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.Base64;
 import org.elasticsearch.rest.RestStatus;
-import org.elasticsearch.test.ElasticsearchIntegrationTest;
 import org.elasticsearch.test.ElasticsearchIntegrationTest.ClusterScope;
-import org.elasticsearch.test.rest.client.http.HttpGetWithEntity;
 import org.elasticsearch.test.rest.client.http.HttpRequestBuilder;
 import org.elasticsearch.test.rest.client.http.HttpResponse;
 import org.junit.Test;
 
-import com.asquera.elasticsearch.plugins.http.HttpBasicServerPlugin;
-
-import java.net.URI;
-import java.net.URISyntaxException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Iterator;
 
 import static org.elasticsearch.test.ElasticsearchIntegrationTest.Scope;
 import static org.hamcrest.Matchers.equalTo;
@@ -46,21 +37,19 @@ import static org.hamcrest.Matchers.equalTo;
  * Test a rest action that sets special response headers
  */
 @ClusterScope(transportClientRatio = 0.0, scope = Scope.SUITE, numDataNodes = 1)
-public class IpAuthenticationIntegrationTest extends ElasticsearchIntegrationTest {
+public class IpAuthenticationIntegrationTest extends HttpBasicServerPluginIntegrationTest {
 
-    protected final String localhost = "127.0.0.1";
     protected final String whitelistedIp = "2.2.2.2";
     protected final String notWhitelistedIp = "3.3.3.3";
     protected final String trustedIp = "4.4.4.4";
 
     @Override
     protected Settings nodeSettings(int nodeOrdinal) {
-        return ImmutableSettings.settingsBuilder()
-                .putArray("http.basic.ipwhitelist", whitelistedIp)
-                .putArray("http.basic.trusted_proxy_chains", trustedIp + "," + localhost)
-                .put("http.basic.xforward", "X-Forwarded-For")
-                .put("plugin.types", HttpBasicServerPlugin.class.getName())
-                .build();
+      return builderWithPlugin()
+        .putArray("http.basic.ipwhitelist", whitelistedIp)
+        .putArray("http.basic.trusted_proxy_chains", trustedIp + "," + localhost)
+        .put("http.basic.xforward", "X-Forwarded-For")
+        .build();
     }
 
     @Test
@@ -70,64 +59,53 @@ public class IpAuthenticationIntegrationTest extends ElasticsearchIntegrationTes
     }
 
     @Test
-    public void localhostClientIsBasicAuthenticated() throws Exception {
-        HttpUriRequest request = httpRequest();
-        String credentials = "admin:admin_pw";
-        request.setHeader("Authorization", "Basic " + Base64.encodeBytes(credentials.getBytes()));
-        CloseableHttpResponse response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.OK.getStatus()));
+    public void clientGoodCredentialsBasicAuthenticationSuceeds() throws Exception {
+        HttpResponse response = requestWithCredentials("admin:admin_pw")
+          .addHeader("X-Forwarded-For", "1.1.1.1" ).execute();
+        assertThat(response.getStatusCode(), equalTo(RestStatus.OK.getStatus()));
     }
 
     @Test
+    public void clientBadCredentialsBasicAuthenticationFails() throws Exception {
+        HttpResponse response = requestWithCredentials("admin:wrong").execute();
+        assertThat(response.getStatusCode()
+            , equalTo(RestStatus.UNAUTHORIZED.getStatus()));
+    }
+    @Test
     public void proxyViaLocalhostIpAuthenticatesWhitelistedClients() throws Exception {
-        HttpUriRequest request = httpRequest();
-        request.setHeader("X-Forwarded-For", whitelistedIp );
-        CloseableHttpResponse response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.OK.getStatus()));
-        request = httpRequest();
-        request.setHeader("X-Forwarded-For", notWhitelistedIp + "," + whitelistedIp);
-        response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.OK.getStatus()));
-        request = httpRequest();
-        request.setHeader("X-Forwarded-For", notWhitelistedIp + "," + whitelistedIp + "," + trustedIp);
-        response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.OK.getStatus()));
+        List<String> whitelists = new ArrayList<String>();
+        whitelists.add(whitelistedIp);
+        whitelists.add(notWhitelistedIp + "," + whitelistedIp);
+        whitelists.add(notWhitelistedIp + "," + whitelistedIp + "," + trustedIp);
+        Iterator<String> iterator = whitelists.iterator();
+        while (iterator.hasNext()) {
+          String next = iterator.next();
+          HttpResponse response = requestWithCredentials("admin:wrong")
+                                  .addHeader("X-Forwarded-For", next)
+                                  .execute();
+          assertThat(response.getStatusCode(), equalTo(RestStatus.OK.getStatus()));
+        }
     }
 
     @Test
     public void proxyViaLocalhostIpUnauthenticatesNonWhitelistedClients() throws Exception {
-        HttpUriRequest request = httpRequest();
-        request.setHeader("X-Forwarded-For", notWhitelistedIp);
-        CloseableHttpResponse response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
-        request = httpRequest();
-        request.setHeader("X-Forwarded-For", whitelistedIp + "," + notWhitelistedIp + "," + trustedIp);
-        response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
-        request = httpRequest();
-        request.setHeader("X-Forwarded-For", "");
-        response = closeableHttpClient().execute(request);
-        assertThat(response.getStatusLine().getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
-    }
-    
-    public static HttpRequestBuilder httpClient() {
-        return new HttpRequestBuilder(HttpClients.createDefault())
-          .host("localhost").port(9200);
-    }
-   
-    public static HttpUriRequest httpRequest() {
-      HttpUriRequest httpUriRequest = null;
-          try {
-            httpUriRequest = new HttpGetWithEntity(new URI("http",
-                  null, "localhost", 9200, "/_status", null, null));
-                } catch (URISyntaxException e) {
-                  throw new IllegalArgumentException(e);
-                }
-      return httpUriRequest;
+        List<String> whitelists = new ArrayList<String>();
+        whitelists.add(notWhitelistedIp);
+        whitelists.add(whitelistedIp + "," + notWhitelistedIp + "," + trustedIp);
+        whitelists.add("");
+        Iterator<String> iterator = whitelists.iterator();
+        while (iterator.hasNext()) {
+          String next = iterator.next();
+          HttpResponse response = requestWithCredentials("admin:wrong")
+                                  .addHeader("X-Forwarded-For", next)
+                                  .execute();
+          assertThat(response.getStatusCode(), equalTo(RestStatus.UNAUTHORIZED.getStatus()));
+        }
     }
 
-    public static CloseableHttpClient closeableHttpClient() {
-      return HttpClients.createDefault();
+    protected HttpRequestBuilder requestWithCredentials(String credentials) throws Exception {
+        return httpClient().path("/_status")
+          .addHeader("Authorization", "Basic " + Base64.encodeBytes(credentials.getBytes()));
     }
-    
+
 }

